server/email_verification.go

```diff
@@ -236,6 +328,12 @@ func handleEmailVerifyFunc(verifiedTpl *template.Template, issuer url.URL, keysF
 			}
 			return
 		}
+
+		qr := cbURL.Query()
+		qr.Set("code", code)
+		qr.Set("state", ses.ClientState)
```

Comment:
NewSession is used to create states for federated services, when dex is acting as a client of an upstream service (e.g. when it's federating to accounts.google.com). Notably you're missing the third argument when calling NewSession, the state dex go from it's client.

Those two states, the one dex got from it's client, and the one dex sends to the upstream service, are not interchangeable.

Here dex is attempting to send the user to dex's client by emulate an Authorization Response. That means redirecting back to dex's client with a code and a state. However state is defined as

> state
>         REQUIRED if the "state" parameter was present in the client
>         authorization request.  The exact value received from the
>         client.

https://tools.ietf.org/html/rfc6749#section-4.1.2

However dex never got a state token from the client, it has no way to fill this field correctly. dex's client MUST always reject a state token that it didn't explicitly create.

Does this make sense?